Code/EnginePlugins/RmlUiPlugin/Components/Implementation/RmlUiCanvas3DComponent.cpp

```diff
@@ -0,0 +1,447 @@
+{
+  m_bClearStaleInput = bClearStaleInput;
+}
+
+void ezRmlUiCanvas3DComponent::SetInteractive(bool bIsInteractive)
+{
+  m_bIsInteractive = bIsInteractive;
+}
+
+void ezRmlUiCanvas3DComponent::OnMsgExtractRenderData(ezMsgExtractRenderData& msg) const
+{
+  if (msg.m_pView->GetCameraUsageHint() == ezCameraUsageHint::MainView || msg.m_pView->GetCameraUsageHint() == ezCameraUsageHint::EditorView)
+  {
+    if (m_pContext != nullptr && m_hTexture.IsValid())
+    {
+      ezResourceLock<ezTexture2DResource> pTexture(m_hTexture, ezResourceAcquireMode::AllowLoadingFallback);
+      if (pTexture.GetAcquireResult() != ezResourceAcquireResult::Final)
+        return;
+
+      ezRmlUi::GetSingleton()->ExtractContext(*m_pContext, pTexture->GetGALTexture());
+    }
+  }
+}
+
+bool ezRmlUiCanvas3DComponent::UpdateTextureAndMaterial()
+{
+  if (m_vSize.x == 0 || m_vSize.y == 0)
+    return false;
+
+  bool bShouldRecreateTexture = !m_hTexture.IsValid();
+
+  if (m_hTexture.IsValid())
+  {
+    ezResourceLock<ezTexture2DResource> pTexture(m_hTexture, ezResourceAcquireMode::AllowLoadingFallback);
+    if (pTexture.GetAcquireResult() != ezResourceAcquireResult::Final)
+      return false;
+
+    bShouldRecreateTexture = pTexture->GetWidth() != m_vSize.x || pTexture->GetHeight() != m_vSize.y;
+  }
+
+  if (bShouldRecreateTexture)
+  {
+    ezTexture2DResourceDescriptor desc;
+    desc.m_DescGAL.m_uiWidth = m_vSize.x;
+    desc.m_DescGAL.m_uiHeight = m_vSize.y;
+    desc.m_DescGAL.m_Format = ezGALResourceFormat::RGBAUByteNormalized;
+    desc.m_DescGAL.m_ResourceAccess.m_bImmutable = false;
+    if (ezMath::IsPowerOf2(m_vSize.x) && ezMath::IsPowerOf2(m_vSize.y))
+    {
+      desc.m_DescGAL.m_uiMipLevelCount = ezMath::Max(ezMath::Log2i(m_vSize.x), ezMath::Log2i(m_vSize.y)) - 2;
+      desc.m_DescGAL.m_bAllowDynamicMipGeneration = true;
+    }
+    
+    ezStringBuilder resourceName = "RmlUiCanvas3DComponent_Texture";
+    resourceName.AppendFormat("_{0}", ezUuid::MakeUuid());
+
+    m_hTexture = ezResourceManager::CreateResource<ezTexture2DResource>(resourceName, std::move(desc));
+
+    m_pContext->SetSize(m_vSize);
+    m_pContext->SetDpiScale(m_fDpiScale);
+
+    m_bRebindTexture = true;
+  }
+
+  if (!m_hMaterial.IsValid())
+  {
+    if (!m_hBaseMaterial.IsValid())
+      return bShouldRecreateTexture;
+
+    ezMaterialResourceDescriptor desc;
+    desc.m_hBaseMaterial = m_hBaseMaterial;
+
+    ezStringBuilder resourceName = "RmlUiCanvas3DComponent_Material";
+    resourceName.AppendFormat("_{0}", ezUuid::MakeUuid());
+
+    m_hMaterial = ezResourceManager::CreateResource<ezMaterialResource>(resourceName, std::move(desc));
+
+    ezMsgSetMeshMaterial msg;
+    msg.m_hMaterial = m_hMaterial;
+    msg.m_uiMaterialSlot = m_uiMaterialIndex;
+    GetOwner()->SendMessage(msg);
+
+    m_bRebindTexture = true;
+  }
+
+  if (m_bRebindTexture)
+  {
+    ezResourceLock<ezMaterialResource> pMaterial(m_hMaterial, ezResourceAcquireMode::AllowLoadingFallback);
```

Comment:
You could also just use BlockTillLoaded since the material is created in code anyways and never load from disk.